### DISCUSSION TITLE
"headers rsa verify failed", but DKIM seems ok

### DISCUSSION THREAD

**Discussion by izydorczyk** [rspamd/rspamd#4782] [General]
I received a message from our servers, the first server with rspamd reported a failed validation and the next server reported a successful validation. This got my attention.
Message path: NEWSLETTER -> MAILHOST -> RELAY -> M365
The NEWSLETTER server created a message, which it forwarded to its MTA and so on. MAILHOST created DKIM signature. The destination mailbox is in M365. On RELAY is rspamd version 3.4 (from the OS repository).

Message headers:
```
DKIM-Signature: v=1; a=rsa-sha256; c=simple/simple; d=newsletter.muni.cz;
	s=inet; t=1705649328;
	bh=ss8/ccYYCOTeErvitwox+7Wn8H8BAaGkoWa0okk2rGY=;
	h=From:Date:Subject:List-Unsubscribe:To:From;
	z=From:=20=3D?utf-8?b?TMOpa2HFmXNrw6E=3D?=3D=20fakulta=20MU=20<no-r
	 eply@newsletter.muni.cz>|Date:=20Fri,=2019=20Jan=202024=2008:28:48
	 =20+0100|Subject:=20=3D?utf-8?b?UMWZZWhsZWQgYWtjw60gbmEgbsOhc2xlZH
	 Vqw61jw60gdMO9ZGVu?=3D|List-Unsubscribe:=0D=0A=20<https://www.med.
	 muni.cz/Redakce/Api/MuniRecipientsApi/Unsubscribe?uco=_reducted_%3d%3d&lang=3Dcs>|To:=20Jan=20Izydorczyk=20<_altered_+IS@_altered_.onmicrosoft.com>;
	b=YkIK1P+TPbGjdxykH7eoH0iJ8ZxTOx1UCNcMM8EDb+Mn/ocTMjAAF7NIxYf+/OjyU
	 KF/1paZg8uOK1SOkVEPCI0lsLvLoMWRab4GRVMv4gc4cJOiKS+dshMrim7LtpeXYAa
	 Wwp0NKPJULtPwI2E644jEgLGp1VYg0NmIVtXqhFtE83JnmW88P8L0pxoDq9bzJ+woc
	 nVybW5O9l5bag7aOJ1JY8S9926GLklFTwap83M2R3TpUI7VR/vhlzkqyDT5svKcGrL
	 sezaCOu53vyT5Gk9/x4EtEzyAPkJSbBsr07o4L6r0+QVme1XECQdQcyZUJEbA2rPPr
	 4yi67hgT0xUmg==
From: =?utf-8?b?TMOpa2HFmXNrw6E=?= fakulta MU <no-reply@newsletter.muni.cz>
Date: Fri, 19 Jan 2024 08:28:48 +0100
Subject: =?utf-8?b?UMWZZWhsZWQgYWtjw60gbmEgbsOhc2xlZHVqw61jw60gdMO9ZGVu?=
List-Unsubscribe:
 <https://www.med.muni.cz/Redakce/Api/MuniRecipientsApi/Unsubscribe?uco=_reducted_%3d%3d&lang=cs>
To: Jan Izydorczyk <_altered_+IS@_altered_.onmicrosoft.com>
``` 
The signature contains a `z=` parameter that contains a copy of the signed headers. Fun part, header from RELAY:
``` 
Authentication-Results-Original: relay.muni.cz;	dkim=fail ("headers rsa verify
 failed") header.d=newsletter.muni.cz header.s=inet header.b=YkIK1P+T;
	spf=pass (relay.muni.cz: domain of no-reply@newsletter.muni.cz designates
 _altered_ as permitted sender) smtp.mailfrom=no-reply@newsletter.muni.cz;
	dmarc=pass (policy=quarantine) header.from=newsletter.muni.cz
``` 
and from M365
``` 
Authentication-Results: spf=pass (sender IP is _altered_)
 smtp.mailfrom=newsletter.muni.cz; dkim=pass (signature was verified)
 header.d=newsletter.muni.cz;dmarc=pass action=none
 header.from=newsletter.muni.cz;compauth=pass reason=100
``` 
First check failed and second pass... part from rspamd log from RELAY:
``` 
2024-01-19 08:28:49 #1881805(rspamd_proxy) <860e02>; dkim; rspamd_dkim_canonize_header: update signature with header (idx=0): From: =?utf-8?b?TMOpa2HFmXNrw6E=?= fakulta MU <no-reply@newsletter.muni.cz>
2024-01-19 08:28:49 #1881805(rspamd_proxy) <860e02>; dkim; rspamd_dkim_canonize_header: update signature with header (idx=0): Date: Fri, 19 Jan 2024 08:28:48 +0100
2024-01-19 08:28:49 #1881805(rspamd_proxy) <860e02>; dkim; rspamd_dkim_canonize_header: update signature with header (idx=0): Subject: =?utf-8?b?UMWZZWhsZWQgYWtjw60gbmEgbsOhc2xlZHVqw61jw60gdMO9ZGVu?=
**2024-01-19 08:28:49 #1881805(rspamd_proxy) <860e02>; dkim; rspamd_dkim_canonize_header: update signature with header (idx=0): List-Unsubscribe:**
2024-01-19 08:28:49 #1881805(rspamd_proxy) <860e02>; dkim; rspamd_dkim_canonize_header: update signature with header (idx=0): To: Jan Izydorczyk <_altered_+IS@_altered_.onmicrosoft.com>
2024-01-19 08:28:49 #1881805(rspamd_proxy) <860e02>; dkim; rspamd_dkim_signature_update: initial update hash with signature part: DKIM-Signature: v=1; a=rsa-sha256; c=simple/simple; d=newsletter.muni.cz;
2024-01-19 08:28:49 #1881805(rspamd_proxy) <860e02>; dkim; rspamd_dkim_check: headers rsa verify failed
``` 
My guess: "List-Unsubscribe:**=0D=0A**=20<https://www.med.muni.cz..."
List-Unsubscribe header, there is a newline before the value. This is confirmed by the "z" parameter in the DKIM signature "=0D=0A".
It looks like the rspamd DKIM check on RELAY ignores the rest of the header on subsequent lines (see logs), whereas M365 does not. Then the DKIM check on RELAY is fail and on M365 pass.
If my assumption is correct, is it possible to modify the behavior of the DKIM check to validate multi-line headers?

**Comment by bmatthewshea** [comment]
Came across your message while investigating the same issue.
Unfortunately, I doubt this will help you much.
I am just putting this here since it may help others..

My situation:
DKIM was passing correctly on incoming from most servers/ESPs (Yahoo/Gmail/et al). But, when I would send from another small Postfix server I host (running amavis - not rspamd), it was showing "headers rsa verify failed" on receiving rspamd server.

For me it seems to have been a Postfix header check/change done on rspamd server - it hides the LAN IP in the first received line for security reasons:

File (/etc/postfix/main.cf):
`header_checks = regexp:/etc/postfix/header_checks`

'header_check' contents (1 line):
`/^Received:.*with ESMTPSA/  REPLACE Received: from localhost`

Once I remove this check from postfix, DKIM passed fine from that server. Oddly enough, DMARC always showed 'passing' entire time.
I suspect this didn't affect all emails because the regex didn't match the "Received:.*with ESMTPSA" line on any except this particular postfix server that was sending. Obviously, this header check needs to be adjusted to be done on only auth'd/outgoing emails (before signing) - not incoming as well.
